### PR TITLE
Hook lambda invoker

### PR DESCRIPTION
**Pull request by ericzbeard**
This PR introduces a Hook that serves as a central point to register Lambda functions that do the actual compliance checks.

### CODE REVIEW

File: config/.pylintrc [hunk 1]
```diff
@@ -396,8 +396,8 @@ preferred-modules=
 [EXCEPTIONS]
 
 # Exceptions that will emit a warning when caught.
-overgeneral-exceptions=BaseException,
-                       Exception
+overgeneral-exceptions=buitlins.BaseException,
```
Comment (mmaeng) [inline config/.pylintrc:399]: I think you have a typo here

**Review by mmaeng — APPROVED** [review]
LGTM

File: hooks/Lambda_Invoker/samples/rain/README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+### module.yaml
+
+This is a rain, module, which is a snippet of CloudFormation that inherits from 
+`AWS::Lambda::Function` to simplify the code in `compliance.yaml`.
+
+### type-config.json
+
+CloudFormation registry extensions must be configured when they are registered. This 
+file is used to tell the hook the Arn of the DynamoDB table that contains a list of
```
Comment (mrinaudo-aws) [inline hooks/Lambda_Invoker/samples/rain/README.md:52]:
```suggestion
file is used to tell the hook the ARN of the DynamoDB table that contains a list of
```

File: hooks/Lambda_Invoker/samples/rain/compliance.yaml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+Resources:
+
+  RegistrationTable:
+    UpdateReplacePolicy: Delete
+    DeletionPolicy: Delete
+    Type: AWS::DynamoDB::Table
+    Metadata:
+      Comment: |
+        This table holds a list of Lambda Arns that will be invoked by the
```
Comment (mrinaudo-aws) [inline hooks/Lambda_Invoker/samples/rain/compliance.yaml:15]:
```suggestion
        This table holds a list of ARNs of Lambda functions that will be invoked by the
```

File: hooks/Lambda_Invoker/samples/rain/module.yaml [hunk 4]
```diff
@@ -0,0 +1,51 @@
+Description: |
+  This is a Rain module that installs a compliance check that will run 
+  on all CloudFormation stacks in the environment where it is installed.
+  This module is invoked from `compliance.yaml` to create a lambda
```
Comment (mrinaudo-aws) [inline hooks/Lambda_Invoker/samples/rain/module.yaml:4]:
```suggestion
  This module is invoked from `compliance.yaml` to create a Lambda 
```

File: hooks/Lambda_Invoker/samples/rain/module.yaml [hunk 5]
```diff
@@ -0,0 +1,51 @@
+Description: |
+  This is a Rain module that installs a compliance check that will run 
+  on all CloudFormation stacks in the environment where it is installed.
+  This module is invoked from `compliance.yaml` to create a lambda 
+  function and add the arn to the registration table. This module
```
Comment (mrinaudo-aws) [inline hooks/Lambda_Invoker/samples/rain/module.yaml:5]:
```suggestion
  function and add the ARN to the registration table. This module 
```

File: hooks/Lambda_Invoker/samples/rain/validate.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
+"""
+This module is a 'shift left' effort to allow a developer to directly invoke
+the compliance lambda functions, to make sure their code is valid before
```
Comment (mrinaudo-aws) [inline hooks/Lambda_Invoker/samples/rain/validate.py:3]:
```suggestion
the compliance Lambda functions, to make sure their code is valid before 
```

File: hooks/Lambda_Invoker/samples/rain/validate.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
+"""
+This module is a 'shift left' effort to allow a developer to directly invoke
+the compliance lambda functions, to make sure their code is valid before 
+trying an actual CloudFormation deployment. It scans the registration table 
+to get the list of lambda function arns, decomposes the template into
```
Comment (mrinaudo-aws) [inline hooks/Lambda_Invoker/samples/rain/validate.py:5]:
```suggestion
to get the list of Lambda function ARNs, decomposes the template into 
```

File: hooks/Lambda_Invoker/samples/rain/validate.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+resources, and submits each of them separately, to mimic the behavior of 
+CloudFormation when it invokes the hook.
+
+This is one of the main reasons for this solution to exist, since CloudFormation 
+hooks do not have an API that can be called to validate templates. By delegating 
+the hook logic out to Lambda functions, we can now call them from anywhere.
+
+One shortcoming of this technique is that we can't know all of the runtime 
+values for functions like GetAtt, for example if you GetAtt a resource Arn, we
+won't know the arn yet, and if your compliance lambda validates a specific arn,
```
Comment (mrinaudo-aws) [inline hooks/Lambda_Invoker/samples/rain/validate.py:15]:
```suggestion
won't know the arn yet, and if your compliance lambda validates a specific ARN, 
```

File: hooks/Lambda_Invoker/samples/rain/validate.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    """
+    args[0]: path to the template
+    args[1]: Arn of the registration table
+    """
+
+    template_path = args[0]
+    table_arn = args[1]
+    table_name = table_arn.split(":table/")[1]
+
+    # Scan the DDB table to get the Lambda Arns
```
Comment (mrinaudo-aws) [inline hooks/Lambda_Invoker/samples/rain/validate.py:36]:
```suggestion
    # Scan the DDB table to get the Lambda ARNs
```

File: hooks/Lambda_Invoker/src/awscommunity_lambda_invoker/invoker.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+                Payload=json.dumps(target),
+            )
+            print(resp)
+            # The invoke function returns 200 if the code raises an exception
+            if "FunctionError" in resp and resp["FunctionError"]:
+                payload = resp["Payload"]
+                j = json.loads(payload.read().decode("utf-8"))
+                print(j)
+                errs.append(j["errorMessage"])
+        except Exception as lamex:
```
Comment (mrinaudo-aws) [inline hooks/Lambda_Invoker/src/awscommunity_lambda_invoker/invoker.py:41]: Can we use an idiomatic name for the variable?

File: hooks/Lambda_Invoker/test/setup.yaml [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+  RegistrationTable:
+    UpdateReplacePolicy: Delete
+    DeletionPolicy: Delete
+    Type: AWS::DynamoDB::Table
+    Metadata:
+      Comment: |
+        This table is created during the testing of the hook. It is also an
+        example of the table that you need to create in order to configure the
+        hook. The table holds a list of Lambda Arns that will be invoked in
```
Comment (mrinaudo-aws) [inline hooks/Lambda_Invoker/test/setup.yaml:17]:
```suggestion
        hook. The table holds a list of ARNs of Lambda functions that will be invoked in
```

**Review by mrinaudo-aws — CHANGES_REQUESTED** [review]
Thank you, @ericzbeard!  I have added suggestions, and feedback.

File: hooks/Lambda_Invoker/README.md [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+itself. Since Lambda functions can be invoked in a variety of ways, it is easier 
+to incorporate those checks into development workflows to move security checks "left"
+so that developers catch problems with their templates sooner.
+
+## Configuration
+
+To use this hook, you must create a DynamoDB table to use for registering your
+compliance Lambda functions. The table can be created with the following
+CloudFormation or manually via the CLI or AWS console. It must have a primary
+key consiting of one attribute, 'lambda\_arn'.
```
[the remark this answers is not in the document]
Comment (ericzbeard) [reply hooks/Lambda_Invoker/README.md:15]: The backslash is necessary for markdown.

File: hooks/Lambda_Invoker/.rpdk-config [hunk 13]
```diff
@@ -0,0 +1,23 @@
+{
+    "artifact_type": "HOOK",
+    "typeName": "AwsCommunity::Lambda::Invoker",
```
[the remark this answers is not in the document]
Comment (ericzbeard) [reply hooks/Lambda_Invoker/.rpdk-config:3]: Maybe `AwsCommunity::Compliance::LambdaInvoker` ?

File: .vscode/settings.json [hunk 14]
```diff
@@ -0,0 +1,6 @@
+{
```
[the remark this answers is not in the document]
Comment (ericzbeard) [reply .vscode/settings.json:1]: We need to use that file specifically or the checks will fail.

File: hooks/Lambda_Invoker/src/awscommunity_lambda_invoker/invoker.py [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+    logger.debug(target)
+    errs = []
+    # Invoke each Lambda Arn and send in the resource properties
+    for arn in lambda_arns:
+        try:
+            resp = lam.invoke(
+                FunctionName=arn.split(":function:")[1],
+                Payload=json.dumps(target),
+            )
+            print(resp)
```
Comment (kddejong) [inline hooks/Lambda_Invoker/src/awscommunity_lambda_invoker/invoker.py:34]: why not logger?

Comment (ericzbeard) [reply hooks/Lambda_Invoker/src/awscommunity_lambda_invoker/invoker.py:34]: Fixing those in the next PR.

File: hooks/Lambda_Invoker/.rpdk-config (hunk 13 above)
[the remark this answers is not in the document]
Comment (ericzbeard) [reply hooks/Lambda_Invoker/.rpdk-config:3]: @kddejong what do you think about the name?

File: hooks/Lambda_Invoker/src/awscommunity_lambda_invoker/invoker.py (hunk 10 above)
Comment (ericzbeard) [reply hooks/Lambda_Invoker/src/awscommunity_lambda_invoker/invoker.py:41]: Doesn't seem important here. I would have made this a one liner but I wanted to see output printed.

**Review by kddejong — APPROVED** [review]

File: hooks/Lambda_Invoker/.rpdk-config (hunk 13 above)
[the remark this answers is not in the document]
Comment (mrinaudo-aws) [reply hooks/Lambda_Invoker/.rpdk-config:3]: Hi @ericzbeard - sorry for delays in responses - I think the `AwsCommunity::Compliance::LambdaInvoker` name you proposed above makes more sense.

**Pull request merged**